src/3d/qgsrulebasedchunkloader_p.h

```diff
@@ -0,0 +1,119 @@
+/***************************************************************************
+  qgsrulebasedchunkloader_p.h
+  --------------------------------------
+  Date                 : November 2019
+  Copyright            : (C) 2019 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSRULEBASEDCHUNKLOADER_H
+#define QGSRULEBASEDCHUNKLOADER_H
+
+///@cond PRIVATE
+
+//
+//  W A R N I N G
+//  -------------
+//
+// This file is not part of the QGIS API.  It exists purely as an
+// implementation detail.  This header file may change from version to
+// version without notice, or even be removed.
+//
+
+#include "qgschunkloader_p.h"
+#include "qgsfeature3dhandler_p.h"
+#include "qgschunkedentity_p.h"
+
+class Qgs3DMapSettings;
+class QgsVectorLayer;
+class QgsVectorLayerFeatureSource;
+class QgsAbstract3DSymbol;
+class QgsFeature3DHandler;
+
+#include "qgsrulebased3drenderer.h"
+
+
+/**
+ * \ingroup 3d
+ * This loader factory is responsible for creation of loaders for individual tiles
+ * of QgsRuleBasedChunkedEntity whenever a new tile is requested by the entity.
+ *
+ * \since QGIS 3.12
+ */
+class QgsRuleBasedChunkLoaderFactory : public QgsChunkLoaderFactory
+{
+  public:
+    //! Constructs the factory
+    QgsRuleBasedChunkLoaderFactory( const Qgs3DMapSettings &map, QgsVectorLayer *vl, QgsRuleBased3DRenderer::Rule *rootRule, int leafLevel );
+    ~QgsRuleBasedChunkLoaderFactory() override;
+
+    //! Creates loader for the given chunk node. Ownership of the returned is passed to the caller.
+    virtual QgsChunkLoader *createChunkLoader( QgsChunkNode *node ) const override;
+
+    const Qgs3DMapSettings &mMap;
```

Comment:
should be these member variables private?

Reply:
They are used directly from QgsRuleBasedChunkLoader so I left them public. This class is implementation detail so it's not in public api anyway...